### PR TITLE
[ao][sparsity] Base class for Data Sparsifier

### PR DESCRIPTION
**Pull request by macandro96**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79535
* #79534
* #79254
* #79253
* #79427
* #79252
* __->__ #79251

Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

Differential Revision: [D37164120](https://our.internmc.facebook.com/intern/diff/D37164120)

### CODE REVIEW

**Review by HDCharles — COMMENTED** [review]
lgtm

File: test/allowlist_for_publicAPI.json [hunk 1]
```diff
@@ -369,6 +369,14 @@
     "Any",
     "List"
   ],
+  "torch.ao.sparsity.experimental.data_sparsifier.base_data_sparsifier": [
```
Comment (albanD) [inline test/allowlist_for_publicAPI.json:372]: No new entry should be added to this file.
In this case, you most likely just want to define `__all__` properly for this submodule.

**Review by albanD — CHANGES_REQUESTED** [review]

File: test/allowlist_for_publicAPI.json (hunk 1 above)
Comment (macandro96) [reply test/allowlist_for_publicAPI.json:372]: These are generic typing packages [List, Any] that are not actually exposed by the submodule (but used by them). Should this still be inside __all__?

Comment (albanD) [reply test/allowlist_for_publicAPI.json:372]: I meant that you should have an `__all__` that contain all the public APIs you want to expose and these should not be there.
if you do that, you won't have to add anything to this allowlist.

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    def get_mask(self, name: str):
+        pass
+
+    def squash_mask(self, *args, **kwargs):
+        pass
+
+    def step(self):
+        pass
+
+    def update_mask(self, name, data, **kwargs):
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:75]: Shouldn't that be an `abstractmethod`?

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        pass
+
+    def load_state_dict(self, state_dict):
+        pass
+
+    def __setstate__(self, state):
+        pass
+
+    def __getstate__(self):
+        pass
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:61]: If these are the placeholders, we can just remove them from this PR, and add them at a later diff

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    def add_data(self, name: str, data, **config):
+        r""" Configures and parametrizes the internal container model with name and data
+
+        Note: The container model is private to the BaseDataSparsifier class
+        """
+        pass
+
+    def get_data(self, name: str):
+        pass
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:49]: Can we add a docstring here. Even if there is no implementation as of now, at least we would get an idea why this method is needed

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    def get_mask(self, name: str):
+        pass
+
+    def squash_mask(self, *args, **kwargs):
+        pass
+
+    def step(self):
+        pass
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:73]: Same "docstring" comment as in the `get_data`

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+
+class BaseDataSparsifier(base_sparsifier.BaseSparsifier):
+    r"""
+    Base Data Sparsifier class for all Data sparsifiers.
+    The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
+    to prepare for sparsification.
+    In this case, mask (and parametrizations) is owned by the class and not by the user.
+    Specifically, the container object inside the class maintains the mask and parametrizations of the input data
+    Args:
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:19]: Space before `Args:`. See [documentation here](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html?highlight=Args#type-annotations) (refer to the "google style")

File: test/allowlist_for_publicAPI.json (hunk 1 above)
Comment (albanD) [reply test/allowlist_for_publicAPI.json:372]: Things like `Any` or `List` are not part of the things your module is exposing publicly (I hope) so these should not be in the `__all__`.

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        - data_list [List[Tuple[str, Any]]]: list of (name, data) tuples to sparsify.
+        Currently - operates on embeddings and torch tensors, parameters
+        Internally, a container module handles the data sparsification.
+
+        - defaults [dict]: default configurations will be attached to the
+            configuration. Only the keys that don't exist in the `config` will
+            be updated.
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:26]: 1. Don't add `-` in front of the parameter name
2. The data type should be in regular parenthesis `(` and `)`
3. The data type should be human readable, not necessarily `typing` instance. For example, use `list of int's` instead of `List[int]`
4. The continuation in the line should be indented.

One option to rewrite this:

```
    Args:
        data_list (list of tuples):
            list of (name, data) tuples to sparsify.
            Currently - operates on embeddings and torch tensors, parameters
            Internally, a container module handles the data sparsification.
        defaults [dict]:
            default configurations will be attached to the
            configuration. Only the keys that don't exist in the `config` will
            be updated.
```

**Review by z-a-f — CHANGES_REQUESTED** [review]

File: test/allowlist_for_publicAPI.json (hunk 1 above)
Comment (z-a-f) [reply test/allowlist_for_publicAPI.json:372]: > 

I think the default behavior in Python is that any imports within a package would be exposed globally to the user. Please, correct me. Here is a basic example that I have, in case it helps understand what I mean: https://colab.research.google.com/drive/1q-p8zfv_p7Fg1AxR4gWveQszNm-ENCZu?usp=sharing

Comment (albanD) [reply test/allowlist_for_publicAPI.json:372]: Yes everything is exposed. but there is nothing we can do about that (unless we force everyone to `del Any, List` etc at the end of every single python file). That's why our public API defition is based on `__all__` when it exists.
Note that when you do `from foo import *`, the `__all__` is what will control what gets imported. So in that sense it does control what gets exposed to the user.

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 2 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:75]: Yep - will add that.

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 6 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:19]: Noted

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 7 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:26]: Noted :)

File: test/allowlist_for_publicAPI.json (hunk 1 above)
Comment (macandro96) [reply test/allowlist_for_publicAPI.json]: added __all__ into the .py file

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 3 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py]: removed them and added them in a later diff

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 2 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:76]: added

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 4 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py]: done.

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 5 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py]: done

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 7 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py]: modified

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 6 above)
Comment (jerryzh168) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:22]: example: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/conv.py#L671

**Review by jerryzh168 — COMMENTED** [review]
LGTM, but I'll let others to accept since I'm not sure what method we should have in the sparsifier

**Review by HDCharles — APPROVED** [review]
lgtm

**Review by z-a-f — APPROVED** [review]
lgtm

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164120).

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164120).

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164120).
